ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp

```diff
@@ -0,0 +1,104 @@
+    TT_FATAL(operation_attributes.stride.size() == 2, "Fold: stride takes 2 elements");
+
+    uint32_t kernel_size_product = 1;
+    uint32_t l = 1;
+    for (uint32_t i = 0; i < 2; ++i) {
+        l *=
+            (((operation_attributes.output_size[i] + 2 * operation_attributes.padding[i] -
+               operation_attributes.dilation[i] * (operation_attributes.kernel_size[i] - 1) - 1) /
+              operation_attributes.stride[i]) +
+             1);
+        kernel_size_product *= operation_attributes.kernel_size[i];
+    }
+
+    TT_FATAL(input_shape.size() == 3, "Fold: Only support 3D input tensor");
+    TT_FATAL(input_shape[-1] == l, "Fold: Invalid input tensor size");
+    TT_FATAL(input_shape[1] % kernel_size_product == 0, "Fold: Invalid input tensor size");
+
+    if (tensor_args.output) {
+        auto& output_shape = tensor_args.output.value().get_legacy_shape();
+        TT_FATAL(output_shape[1] == input_shape[1] / kernel_size_product, "Fold: Invalid output tensor size");
+    }
+}
+
+MorehFoldOperation::program_factory_t MorehFoldOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return ProgramFactory{};
+};
+
+void MorehFoldOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+void MorehFoldOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehFoldOperation::shape_return_value_t MorehFoldOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.input.get_logical_shape();
+    uint32_t kernel_size_product = operation_attributes.kernel_size[0] * operation_attributes.kernel_size[1];
+    uint32_t N = input_tensor_shape[0];
+    uint32_t C = input_tensor_shape[1] / kernel_size_product;
+    auto output_shape = ttnn::SimpleShape(
+        std::vector<uint32_t>{N, C, operation_attributes.output_size[0], operation_attributes.output_size[1]});
+    return output_shape;
+};
+
+MorehFoldOperation::tensor_return_value_t MorehFoldOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_shape = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.input.get_dtype();
+    auto layout = tensor_args.input.get_layout();
+    auto device = tensor_args.input.device();
+    if (tensor_args.output.has_value()) {
+        return tensor_args.output.value();
+    }
+    return create_device_tensor(output_shape, dtype, layout, device, operation_attributes.output_memory_config);
+}
```

Comment:
Optimization suggestion: Move the check for `tensor_args.output.has_value()` to the top to avoid unnecessary computations. 

Also, remove intermediate variable declarations to streamline the code.

Reply:
Resolved in last commit.